### PR TITLE
Move IsStream type class definition out of the low level StreamK module

### PR DESCRIPTION
**Pull request by harendra-kumar**
## Description

Move `IsStream` definition out of the low level `StreamK` module and push it up in a higher level IsStream/Type module. This is a step towards creating a streamly-core package devoid of any non-boot library dependencies, as mentioned in #533 . Also, this will facilitate moving to monomorphic stream types and modules in future, currently we use polymorphic streams based on IsStream type class.

The IsStream type class contains MonadAsync type in the consM method. This refactor allows lifting MonadAsync, therefore, dependency on monad-control out of the low level modules.

The main changes are in StreamK/Type.hs, StreamK.hs, IsStream/Type.hs, Stream/Serial.hs, the rest of the changes are mostly a fallout of these changes. Need to review these files and a few other files with important changes, the rest of the changes can be glanced over.

## Things to check

* Performance remains the same
* Documentation is not lost
* Exposed APIs do not change - cabal-diff
* Concurrent APIs remain concurrent

## Performance Notes

The `foldrS` benchmark has regressed by around 100 times, but the original perf numbers were good because of
a trivial foldr/build fusion. If we see `foldrSMap` which actually performs a map
using `foldrS` then it has improved by 2x.

```
Benchmark  default(0)(ns) default(1)(ns)
All.Prelude.Serial/o-1-space.mapping.foldrS 37636.90 3002730.00
All.Prelude.Serial/o-1-space.mapping.foldrSMap 6872930.00 2895490.00
```

`iterateState` benchmark now requires a SPECIALIZE pragma to specialize.

Other degradations, but the core looks the same for all these cases, may want to verify once again:

```
ParserD.serialWith 50% degradation, many
Parser serialWith/split_ 50%
Array.Foreign writeN/writeLastN/fromList 50% deg
```

Other than these there are no significant regressions, there are many improvements though.

### CODE REVIEW

File: src/Streamly/Internal/Data/Stream/IsStream/Type.hs [hunk 1]
```diff
@@ -0,0 +1,698 @@
...
+-- parameterized by that stream.
+--
+-- XXX Since t is coercible we should be able to coerce k
+-- mkStream k = fromStream $ MkStream $ coerce k
+--
+-- | Build a stream from an 'SVar', a stop continuation, a singleton stream
+-- continuation and a yield continuation.
+{-# INLINE_EARLY mkStream #-}
+mkStream :: IsStream t
+    => (forall r. State K.Stream m a
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/IsStream/Type.hs:260]: I remember there being an issue or a comment about fixing `K.Stream` in `State`. No comment. Just stating it here.

File: src/Streamly/Internal/Data/Stream/StreamD/Eliminate.hs [hunk 2]
```diff
@@ -228,19 +228,6 @@ parse_ (PRD.Parser pstep initial extract) stream@(Stream step state) = do
 -- Specialized Folds
 ------------------------------------------------------------------------------
 
--- | Run a streaming composition, discard the results.
-{-# INLINE_LATE drain #-}
-drain :: Monad m => Stream m a -> m ()
--- drain = foldrM (\_ xs -> xs) (return ())
-drain (Stream step state) = go SPEC state
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/StreamD/Eliminate.hs:235]: Why was this removed/moved?

File: benchmark/Streamly/Benchmark/Data/Stream/StreamK.hs [hunk 3]
```diff
@@ -471,7 +480,7 @@ sourceConcatMapId val n =
 {-# INLINE concatMapBySerial #-}
 concatMapBySerial :: Int -> Int -> Int -> IO ()
 concatMapBySerial outer inner n =
-    S.drain $ S.concatMapBy S.serial
+    S.drain $ S.concatMapWith S.serial
```
Comment (adithyaov) [inline benchmark/Streamly/Benchmark/Data/Stream/StreamK.hs:483]: Are `concatMapWith` and `concatMapBy` same?

**Review by adithyaov — APPROVED** [review]
I went through them. Looks good. Did you tally and make sure you've not removed any API by mistake? You're using some tool I believe.

File: benchmark/Streamly/Benchmark/Data/Stream/StreamK.hs (hunk 3 above)
Comment (harendra-kumar) [reply benchmark/Streamly/Benchmark/Data/Stream/StreamK.hs:483]: concatMapBy is older name, later renamed to concatMapWith.

File: src/Streamly/Internal/Data/Stream/StreamD/Eliminate.hs (hunk 2 above)
Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Stream/StreamD/Eliminate.hs:235]: It is required in low level modules like instances of SerialT and other types.

**Comment by harendra-kumar** [comment]
`cabal-diff` is the tool. There is one change to the exposed API, I will fix that. There are several changes to Internal APIs. I am attaching the diff.

[API diff of master with 0.8.0](https://github.com/composewell/streamly/files/7149354/diff.txt)

File: bin/test.sh [hunk 4]
```diff
@@ -59,14 +59,14 @@ do
     # options with arguments
     --targets) shift; TARGETS=$1; shift ;;
     --with-compiler) shift; CABAL_WITH_COMPILER=$1; shift ;;
-    --cabal-build-options) shift; CABAL_BUILD_OPTIONS+=$1; shift ;;
-    --hpc-report-options) shift; HPC_REPORT_OPTIONS=$1; shift ;;
-    --rtsopts) shift; RTS_OPTIONS=$1; shift ;;
+    --cabal-build-options) shift; CABAL_BUILD_OPTIONS+=" $1"; shift ;;
```
Comment (adithyaov) [inline bin/test.sh:62]: I fixed this in another change. I'll undo that.

**Comment by harendra-kumar** [comment]
I am going to squash and merge it.
